### ISSUE TITLE
Submerge hides units over bridges

### ISSUE THREAD

**Issue by DisherProject** [wesnoth/wesnoth#3605]
![screenshot 1](https://user-images.githubusercontent.com/29802612/46586788-b28e5900-ca83-11e8-9f06-eb803feb0837.png)

**Comment by gfgtdf** [comment]
sounds like the expected behaviour to me, although it would probalby be better if the unit was drawn under the bridge which wensoth currently cannot do afaik.

**Comment by Wedge009** [comment]
I would have thought this is expected behaviour, if the definition of submerge (at least outside Wesnoth) involves being _under_ water. Is the issue because bridges are considered a combination of flat and water? Could submerge be modified to only apply in deep water and deep water alone? Would that have adverse impacts on anything else? (Combination with swamps, perhaps?) I'm not too sure how the terrain handling works.

**Comment by GregoryLundberg** [comment]
To my mind, if the unit entered the bridge tile(s) from land, then they are on land and if they entered from water, they are on water. This status is set upon moving to the first bridge tile and cleared/lost upon moving onto the next non-bridge tile. You probably could do it in WML as a proof-of-concept but it probably should be in C++ or Lua.

**Comment by Wedge009** [comment]
I was thinking something similar, but I believe it's beyond the scope of what Wesnoth currently handles. Maybe it's an option for the 'Wesnoth 2 rule-set'?

**Comment by sevu** [comment]
How about disabling submerge under bridges (by exycluding the terrain codes in the submerge macro)?

Mainline submerge is only used in deep water, bridges are usually on shallow water, so there wouldn't be gamplay considerations.

**Comment by jostephd** [comment]
I don't think we should change the behavior of submerge to accommodate the UI. There are valid gameplay reasons to have a bridge over deep water (for example, a drawbridge over a moat), I'd like to allow that gameplay even if we display it not ideally.

**Comment by GregoryLundberg** [comment]
How about having bridges show the units and openning a ticket "Bridges expose submerged units."? /s

**Comment by ln-zookeeper** [comment]
I don't think it's possible to fix all the involved inconsistencies. Either you let the lich enjoy the defense of flat but somehow still be able to submerge in the deep water below (current behavior), or you prevent the lich from choosing to submerge in the deep water that it should obviously be able to jump in (implicitly proposed behavior). Neither is particularly less arbitrary than the other, but at least the current way allows a unit to enjoy all the benefits of the tile, which is simpler and fairer than a hardcoded tradeoff.

Realism-wise, you can always think of the lich hiding in the deep water below the bridge so that it can't be seen from a distance, and then climbing up to ambush the passerby. In this case, the only complaint left would be that the bridge should not height-adjust submerging units. Also, in theory one could even have an ambush animation where the unit slides up from the water (although possibly not possible currently) to visually establish said interpretation.

**Comment by Wedge009** [comment]
I wonder how 2.0 would work out this interesting boundary case.

**Comment by doofus-01** [comment]
This isn't a terrain issue, it's just the filtering in the ability.
```
#define ABILITY_SUBMERGE
    ...
            [filter_location]
                terrain=Wo*^*
            [/filter_location]
    ...
```
Whatever conceptual problem there is with lichs not being able to jump in the water underneath is shared with just about anything in Wesnoth.  If it's really important to someone, they can script it into their scenario.  If enough people think there is something wrong with how bridges are handled, maybe a decision can be made to adjust the aliasof/def_alias keys.  But as it is, there is no terrain bug.

**Comment by ProditorMagnus** [comment]
I plan to review the hiding terrains soon https://github.com/inferno8/wesnoth-Era_of_Magic/issues/70

**Comment by CelticMinstrel** [comment]
> This isn't a terrain issue, it's just the filtering in the ability.

I disagree. The correct solution is to draw the unit under the bridge, without any height adjustment due to it being a bridge, which currently isn't supported if I recall correctly.

**Comment by ProditorMagnus** [comment]
It is not that simple. In case of land unit, bridge provides better movement and defense, so unit should be drawn there, and submerge should be disabled (unless unit is resting).

**Comment by CelticMinstrel** [comment]
I'm not so sure about that… even if it provides better movement and defense, the submerge seems like a great trade-off for the poor movement and defense.

**Comment by doofus-01** [comment]
That's a decision that could be made and the terrain changed to reflect that.  It's a decision I wouldn't agree with, but it could be done.  There is no bug here, except for the submerge filtering.

If you want terrains to have different rules for different units, that sounds confusing.  I hope that doesn't become a thing.

**Comment by Wedge009** [comment]
> This isn't a terrain issue...

The way I usually consider labels isn't necessarily to say, in this case, that it's a terrain-specific _bug_ but to note that the issue does relate at least in some way to terrains, broadly. I find it helpful when searching for issues via tags.

**Comment by CelticMinstrel** [comment]
> If you want terrains to have different rules for different units, that sounds confusing. I hope that doesn't become a thing.

I'm not quite sure what you mean by "different rules for different units"…

**Comment by doofus-01** [comment]
> I'm not quite sure what you mean by "different rules for different units"…

A spearman walks onto a bridge.  A lich walks onto a bridge.  You want the deep water rules to apply for the lich.

> The way I usually consider labels isn't necessarily to say, in this case, that it's a terrain-specific bug but to note that the issue does relate at least in some way to terrains, broadly. I find it helpful when searching for issues via tags.

Fair enough, just not how I read it, I guess.  I was filtering for terrain bugs, and in this case, I don't think there is anything sane to be done for this, as far as terrain or terrain graphics.

**Comment by Wedge009** [comment]
> You want the deep water rules to apply for the lich.

Isn't it the case that the deep water rules should apply to units with the submerge ability, not the lich unit type specifically?

I grant this is a curious case that probably isn't going to be resolved easily.

**Comment by Wedge009** [comment]
While jarring from a real-world perspective, this incongruence seems to be a product of Wesnoth's design choice. Reviewing the Help text:

https://github.com/wesnoth/wesnoth/blob/c284a080e3d7c93652598bb71b1bad95620263f6/data/core/help.cfg#L576

'One notable exception is bridge terrains... Fords are easily passable to both merfolk and humans — all units moving on a ford enjoy the best defence and best movement out of flat and shallow water...'

I'm probably just repeating what has been said above, but I found this while reviewing the aforementioned text.

**Comment by doofus-01** [comment]
It's not that jarring.  If the Lich, because it had submerge ability, got deep-water rules to apply on a bridge, was then ambushed on a bridge, it would get the deep water or flat terrains?  Is it a ninja lich that can jump to whatever is best for the situation?  Or does it just walk across the bridge like a spearman?  The skeleton movement in deep water sucks, there's no reason to think they are SEAL Team Six.
